BossMod/QuestBattle/QuestBattleDirector.cs

```diff
@@ -0,0 +1,375 @@
+}
+
+public sealed class QuestBattleDirector : IDisposable
+{
+
+    public readonly WorldState World;
+    private readonly BossModuleManager bmm;
+    private readonly EventSubscriptions _subscriptions;
+    private readonly QuestBattleConfig _config;
+
+    public readonly record struct NavigationWaypoint(Vector3 Position, bool SpecifiedInPath);
+
+    public List<NavigationWaypoint> CurrentWaypoints { get; private set; } = [];
+    public QuestBattle? CurrentModule { get; private set; }
+    public QuestObjective? CurrentObjective { get; private set; }
+
+    public int CurrentObjectiveNavigationProgress { get; private set; }
+    public List<Waypoint> CurrentConnections { get; private set; } = [];
+
+    public bool Paused;
+
+    public Event<QuestBattle> QuestActivated = new();
+    public Event<QuestObjective> ObjectiveChanged = new();
+    public Event<QuestObjective> ObjectiveCleared = new();
+
+    public const float Tolerance = 0.25f;
+
+    private readonly ICallGateSubscriber<Vector3, Vector3, bool, Task<List<Vector3>>?> _pathfind;
+    private readonly ICallGateSubscriber<bool> _isMeshReady;
+
+    private bool _combatFlag;
+
+    private static void Log(string msg) => Service.Log($"[QBD] {msg}");
+
+    public bool Enabled => _config.Enabled && World.CurrentCFCID != 0 && World.Party.WithoutSlot(includeDead: true).Count(x => x.Type == ActorType.Player) == 1;
+
+    public QuestBattleDirector(WorldState ws, BossModuleManager bmm)
+    {
+        World = ws;
+        _config = Service.Config.Get<QuestBattleConfig>();
+        this.bmm = bmm;
+
+        _subscriptions = new(
+#if DEBUG
+            ws.Actors.StatusGain.Subscribe((a, i) =>
+            {
+                if (a.OID == 0)
+                    return;
+                Log($"{a} gain {a.Statuses[i]}");
+            }),
+            ws.Actors.StatusGain.Subscribe((a, i) =>
+            {
+                if (a.OID == 0)
+                    return;
+                Log($"{a} lose {a.Statuses[i]}");
+            }),
+            ws.Actors.ModelStateChanged.Subscribe(a =>
+            {
+                if (a.OID == 0)
+                    return;
+                Log($"{a} model state: {a.ModelState}");
+            }),
+            ws.Actors.EventStateChanged.Subscribe(a =>
+            {
+                if (a.OID == 0)
+                    return;
+                Log($"{a} event state: {a.EventState}");
+            }),
+            ws.DirectorUpdate.Subscribe(diru =>
+            {
+                Log($"Director update: U:0x{diru.UpdateID:X} P1:0x{diru.Param1:X} P2:0x{diru.Param2:X} P3:0x{diru.Param3:X} P4:0x{diru.Param4:X}");
+            }),
+            ws.Actors.EventObjectStateChange.Subscribe((act, u) =>
+            {
+                Log($"EObjState: {act} => 0x{u:X}");
+            }),
+            ws.Actors.EventObjectAnimation.Subscribe((act, p1, p2) =>
+            {
+                Log($"EObjAnim: {act}, {p1}, {p2}");
+            }),
+#endif
```

Comment:
Can we remove all that? All this info should be available in replays.

Reply:
yessir